xapian-applications/omega/handler_pdf.cc

```diff
@@ -0,0 +1,93 @@
+/** @file handler_pdf.cc
+ * @brief @brief Extract text and metadata using poopler.
+ */
+/* Copyright (C) 2011 Olly Betts
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+#include <config.h>
+
+#include <xapian.h>
+#include "handler_pdf.h"
+#include <iostream>
+#include <poppler-document.h>
+#include <poppler-page.h>
+#include <poppler-version.h>
+
+using namespace std;
+using namespace poppler;
+
+static string
+clear_text(const ustring & x)
+{
+    byte_array buf = x.to_utf8();
+    string text(buf.data(), buf.size());
+    int i, j, sz = text.length();
+
+    for (j = i = 0; i < sz; ++i)
+	if (!isspace(text[i]) || (i && !isspace(text[i - 1])))
+	    text[j++] = text[i];
+    text.resize(j);
+
+    return text;
+}
+
+bool
+extract(const string & filename,
+	string & dump,
+	string & title,
+	string & keywords,
+	string & author)
+{
+
+    try {
+
+	document * doc = document::load_from_file(filename, "", "");
+
+	if (!doc) {
+	    cerr << "Poppler Error: Failed to read pdf file" << endl;
+	    return false;
+	}
+
+#if (POPPLER_VERSION_MAJOR>0 || POPPLER_VERSION_MINOR>=46)
+	    author = clear_text(doc->get_author());
+	    title = clear_text(doc->get_title());
+	    keywords = clear_text(doc->get_keywords());
+	    cout << "author: <<" << author << ">>" << endl;
+	    cout << "title: <<" << author << ">>" << endl;
+	    cout << "keywords: <<" << author << ">>" << endl;
+#else
+	    author = "";
+	    title = "";
+	    keywords = "";
+#endif
```

Comment:
I am having problems with Poppler version (packages.ubuntu.com/xenial/libpoppler-cpp-dev) and Travis.
Travis uses version 0.41 (Feb 16, 2016), but I need at least 0.46 (Jul 25, 2016) to be able to use method get_author, get_title and get_keywords.
Xenial use 0.41, but Bionic and newer uses poppler version >=0.62.

However, in 0.41 I have a method metadata (it returns a `ustring` with xml format) from which I can extract some information (title, author) but it doesn't as well as the other method. In some cases metadata is empty and get_author and get_title works ok.

I also found this _ppa:opencpu/poppler_ [https://launchpad.net/~opencpu/+archive/ubuntu/poppler](url). I am not sure if it is useful.

Reply:
Can you get the author &c information using `::info_key()`? It looks from the source code like this will work, and is supported back to 0.41 or earlier.

Reply:
Yes, you are right. I didn't notice that